// @herodotus_dev/mmr-core@1.1.17/lib/trees-database.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InStoreCounter = exports.InStoreTable = exports.TreesDatabase = void 0;
const types_1 = require("./types");
const uuid_1 = require("uuid");
class TreesDatabase {
    store;
    mmrUuid;
    leavesCount;
    elementsCount;
    hashes;
    rootHash;
    constructor(store, mmrUuid) {
        this.store = store;
        mmrUuid ? (this.mmrUuid = mmrUuid) : (this.mmrUuid = (0, uuid_1.v4)());
        this.leavesCount = new InStoreCounter(this.store, `${this.mmrUuid}:${types_1.TREE_METADATA_KEYS.LEAF_COUNT}`);
        this.elementsCount = new InStoreCounter(this.store, `${this.mmrUuid}:${types_1.TREE_METADATA_KEYS.ELEMENT_COUNT}`);
        this.rootHash = new InStoreTable(this.store, `${this.mmrUuid}:${types_1.TREE_METADATA_KEYS.ROOT_HASH}`);
        this.hashes = new InStoreTable(this.store, `${this.mmrUuid}:hashes:`);
    }
}
exports.TreesDatabase = TreesDatabase;
class InStoreTable {
    store;
    key;
    constructor(store, key) {
        this.store = store;
        this.key = key;
    }
    getFullKey(suffix) {
        return this.key + (suffix ?? "").toString() || "";
    }
    async get(suffix) {
        return this.store.get(this.getFullKey(suffix));
    }
    async getMany(suffixes) {
        const keys = suffixes.map((suffix) => this.getFullKey(suffix));
        const keyless = new Map();
        (await this.store.getMany(keys)).forEach((value, key) => keyless.set(key.split(":").slice(2).join(":"), value));
        return keyless;
    }
    async set(value, suffix) {
        return this.store.set(this.getFullKey(suffix), value);
    }
}
exports.InStoreTable = InStoreTable;
class InStoreCounter {
    store;
    key;
    constructor(store, key) {
        this.store = store;
        this.key = key;
    }
    async get() {
        const count = await this.store.get(this.key);
        return count ? parseInt(count) : 0;
    }
    async set(count) {
        if (isNaN(count))
            throw new Error("Count is not a number");
        await this.store.set(this.key, count.toString());
    }
    async increment() {
        const count = await this.get();
        const newCount = count + 1;
        await this.set(newCount);
        return newCount;
    }
}
exports.InStoreCounter = InStoreCounter;
